Feature - Adding sorting for name, interview date and priority

laughingkid-sg opened the pull request:
# Completed sorting feature. 

- Implemented sorting logic in model
- Developed as an addon to list
- Allow sorting by the following parameters - name, interview date, priority, status
- Allow sorting in both descending and ascending 
- Updated test case to ensure no errors 
- Updated user guide
- Added Javadoc

Fix #109

laughingkid-sg committed:
Update Model
Updated Model and Model Manager class to allow modification of data in
memory

laughingkid-sg committed:
Add Comparators
Added 3 comparators (name, interview slot, priority) to allow sorting
of applications via the 3 fields

laughingkid-sg committed:
Add command parser for list command
- Added command parser for list to handle different sorting arguments
- Updated InternApplyParser to handle arguments for list command

laughingkid-sg committed:
Update list command
- Added sortingComparator and orderBy field
- Updated execute to sort model

laughingkid-sg committed:
Add getter for InterviewSlot and PriorityTag fields

laughingkid-sg committed:
Merge branch 'master' into 109-branch-sorting

laughingkid-sg committed:
Merge branch 'master' into 109-branch-sorting

laughingkid-sg committed:
Update Priority Comparator

Fixed priority comparator comparing wrong values.

laughingkid-sg committed:
Update sorting logic

- Removed sorting using FilteredApplicationList class
- Added sorting methods using InternApplyMemory class

laughingkid-sg committed:
Update PriorityTag getter

laughingkid-sg committed:
Update ListCommand test cases

lchokhoe commented on src/main/java/seedu/address/logic/sort/InterviewSlotComparator.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public int compare(Application o1, Application o2) {
+        return o1.getInterviewSlot().getValue().compareTo(o2.getInterviewSlot().getValue());
```
LGTM. But maybe you can consider adding in a short JavaDoc for this method?

lchokhoe replied:
Nice summarized JavaDoc!

lchokhoe commented on src/main/java/seedu/address/logic/sort/NameComparator.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public int compare(Application o1, Application o2) {
+        return o1.getName().toString().compareTo(o2.getName().toString());
```
LGTM. But maybe you can consider adding in a short JavaDoc for this method?

lchokhoe replied:
Nice summarized JavaDoc!

lchokhoe commented on src/main/java/seedu/address/logic/sort/PriorityComparator.java:
```diff
@@ -0,0 +1,42 @@
+    private static final String PRIORITY_LOW = "[" + PriorityTagType.LOW + "]";
+
+    @Override
+    public int compare(Application o1, Application o2) {
```
For this there was an alternative of converting the TagName to a corresponding Integer that we discussed.

Alternatively, to make the logic more understandable I think you can include some comments. And also a short summary of what is being compared as a JavaDoc?

lchokhoe replied:
Nice summarized JavaDoc!

lchokhoe commented on src/main/java/seedu/address/logic/commands/ListCommand.java:
```diff
@@ -12,12 +19,46 @@
+                + PriorityComparator.COMMAND_WORD.toUpperCase() + "] "
+            + "[order " + COMMAND_ORDER_WORD_ASCENDING + "|" + COMMAND_ORDER_WORD_DESCENDING + "] \n"
+            + "Example: " + COMMAND_WORD + " INTERVIEW " + COMMAND_ORDER_WORD_DESCENDING + "\n"
+            + "Note: Using list without parameters will revert to the last sorted.";
```
Good use of String's to avoid creating "magic numbers". Also aligns with the way logic has been coded for this software.

Nice!

lchokhoe commented on src/main/java/seedu/address/logic/commands/ListCommand.java:
```diff
@@ -12,12 +19,46 @@
+     */
+    public ListCommand() {
+        sortingComparator = Optional.empty();
+        orderBy = COMMAND_ORDER_WORD_ASCENDING;
```
Does this mean that by Default it will always display in ascending order? If so maybe you can indicate some where in the JavaDoc as well

laughingkid-sg replied:
just placeholder, updated javadoc

lchokhoe replied:
Nice summarized JavaDoc!

lchokhoe reviewed (commented):
Just have some comments about documentation (mostly). Other then that LGTM.

laughingkid-sg committed:
Merge branch 'master' into 109-branch-sorting

laughingkid-sg committed:
Add sorting logic for ApplicationStatus

- Added getter for ApplicationStatusTag
- Added comparator for ApplicationStatusTag

laughingkid-sg committed:
Merge branch 'master' into 109-branch-sorting

laughingkid-sg committed:
Update test cases

laughingkid-sg committed:
Add sorting by ApplicationStatus

- Added sorting parser for application status tag
- Updated list success message
- Change sorting order for application status tag

laughingkid-sg committed:
Add comments for new methods and constructors

laughingkid-sg committed:
Update UserGuide.md

Updated list feature to accommodate sorting parameters.

laughingkid-sg committed:
Update UserGuide.md

Update list command in summary

laughingkid-sg committed:
Merge branch 'master' into 109-branch-sorting

lchokhoe commented on src/main/java/seedu/address/logic/parser/ListCommandParser.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public class ListCommandParser implements Parser<ListCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the ListCommand
+     * and returns a ListCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
```
Nice summarized JavaDoc!

lchokhoe commented on src/main/java/seedu/address/logic/sort/ApplicationStatusComparator.java:
```diff
@@ -0,0 +1,53 @@
+        return getStatusRanking(o1.getApplicationStatusTag()) - getStatusRanking(o2.getApplicationStatusTag());
+    }
+
+    /**
+     * Compares the two application's status tag field. Returns a negative integer, zero, or a positive integer as
+     * the first argument is less than, equal to, or greater than the second base on alphanumeric order.
+     * The default order from ascending to descend is as follows ACCEPTED, REJECTED, INTERVIEWED, APPLIE and
+     * NOT_APPLIED.
+     * An empty tag will be considered as the lowest ranking, ranked below NOT_APPLIED.
+     * */
```
Nice summarized JavaDoc!

lchokhoe commented on src/main/java/seedu/address/model/application/Application.java:
```diff
@@ -79,6 +81,24 @@ public Details getDetails() {
         return details;
     }
 
+    public Optional<Tag> getPriorityTag() {
+        for (Tag tag: tags) {
+            if (tag.tagType.equals(TagType.PRIORITY)) {
+                return Optional.of(tag);
+            }
+        }
+        return Optional.empty();
+    }
+
+    public Optional<Tag> getApplicationStatusTag() {
+        for (Tag tag: tags) {
+            if (tag.tagType.equals(TagType.APPLICATION_STATUS)) {
+                return Optional.of(tag);
+            }
+        }
+        return Optional.empty();
+    }
+
```
Like your use of Optional here. Also good use of the TagType to look through the Set<Tag> tags!

lchokhoe reviewed (approved):
LGTM. Will merge now.
